File: C++ Folder/OOP/mergeshort.cpp
#include <bits/stdc++.h>
using namespace std;
//MARK: Binary searchf
void targetfound(int si, int en, int arr[], int target)
{
    int targetidx = -1;
    int start = si, end = en;
    int mid = start + (end - start) / 2;
    while (start <= end)
    {
        mid = start + (end - start) / 2;

        if (arr[mid] == target)
        {
            targetidx = mid;
            break;
        }
        else if (arr[mid] > target)
        {
            end = mid - 1;
        }
        else
        {
            start = mid + 1;
        }
    }
    if(arr[targetidx] == target)
    {
    cout << "Target index is: " << targetidx << endl;
    }
    
}
//MARK: Merge sort sorting algorithm
void conqure(int arr[], int si,int mid, int en){

    int merge[en-si+1];
    int idx1= si,idx2= mid+1,x=0;
    while(idx1<=mid && idx2<=en){
        if(arr[idx1]<arr[idx2]){
            merge[x] = arr[idx1];
            x++; idx1++;
        }
        else{
            merge[x] = arr[idx2];
            x++; idx2++;
        }
    }
    while(idx1<=mid){
        merge[x] = arr[idx1];
        x++; idx1++;
    }
    while(idx2<=en){
        merge[x] = arr[idx2];
        x++; idx2++;
    }
    for (int i = 0; i < en-si+1; i++)
    {
        arr[si+i] = merge[i];
    }
    //printing array
    for (int i = 0; i < en-si+1; i++)
    {
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    targetfound(si,en, merge, 13);
}
//MARK: Divide function
void divide(int arr[], int si, int en)
{
    if(si>=en){
        return;
    }
    int mid = si + (en - si)/2;
    divide(arr,si,mid);
    divide(arr,mid+1,en);
    conqure(arr,si,mid,en);
    
}
//MARK: Main function
int main()
{
    int arr[] = {12, 11, 13, 5, 6, 7};
    int sizeofarr = sizeof(arr)/sizeof(arr[0]);
    //calling function
    divide(arr,0,sizeofarr-1);
    cout<<endl;
    


return 0;
}

File: C++ Folder/OOP/Others/mergeShort.swift
func mergeSort(numbers: inout [Int], start: Int, end: Int) {
    if start >= end {
        return
    }
    
    let mid = start + (end - start) / 2
    mergeSort(numbers: &numbers, start: start, end: mid)
    mergeSort(numbers: &numbers, start: mid + 1, end: end)
    merge(numbers: &numbers, start: start, mid: mid, end: end)
}

func merge(numbers: inout [Int], start: Int, mid: Int, end: Int) {
    var temp: [Int] = []
    var idx1 = start, idx2 = mid + 1
    
    while idx1 <= mid && idx2 <= end {
        if numbers[idx1] < numbers[idx2] {
            temp.append(numbers[idx1])
            idx1 += 1
        } else {
            temp.append(numbers[idx2])
            idx2 += 1
        }
    }
    
    while idx1 <= mid {
        temp.append(numbers[idx1])
        idx1 += 1
    }
    
    while idx2 <= end {
        temp.append(numbers[idx2])
        idx2 += 1
    }
    
    for i in start...end {
        numbers[i] = temp[i - start]
    }
}

func main() {
    var numbers = [3, 4, 2, 77, 43, 44, 23, 34, 64, 21, 33, 44, 55, 66, 77, 88, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120]
    let start = 0, end = numbers.count - 1
    mergeSort(numbers: &numbers, start: start, end: end)
    
    for number in numbers {
        print("\(number) ", terminator: "")
    }
}

main()